tests/integration/test_data_store_mgr.py

```diff
@@ -313,12 +344,93 @@ async def test_update_data_structure(harness):
     assert TASK_STATUS_FAILED in set(collect_states(data, FAMILY_PROXIES))
     # state totals changed
     assert TASK_STATUS_FAILED in data[WORKFLOW].state_totals
-    # Shows pruning worked
-    # TODO: fixme
-    # https://github.com/cylc/cylc-flow/issues/4175#issuecomment-1025666413
-    # assert len({t.is_held for t in data[TASK_PROXIES].values()}) == 1
 
 
+async def test_prune_data_store(harness_prune):
+    """Test prune_data_store. This method will expand and reduce the data-store
+    to invoke pruning."""
+    schd, data = harness_prune
+    w_id = schd.data_store_mgr.workflow_id
+    schd.data_store_mgr.data[w_id] = data
+    schd.pool.hold_tasks(['*'])
+    await schd.update_data_structure()
+    assert len({t.is_held for t in data[TASK_PROXIES].values()}) == 2
+    # Window size reduction to invoke pruning
+    schd.data_store_mgr.set_graph_window_extent(0)
+    schd.data_store_mgr.update_data_structure()
+    assert len({
+        t.is_held
+        for t in data[TASK_PROXIES].values() 
+        if t.graph_depth == 0
+    }) == 1
+
+    # Test rapid addition and removal
+    assert len({
+        f.name
+        for f in schd.data_store_mgr.added[FAMILY_PROXIES].values()
+        if f.name == 'BAR'
+    }) == 0
+    for itask in schd.pool.get_tasks():
+        schd.pool.spawn_on_output(itask, TASK_OUTPUT_SUCCEEDED)
+    assert len({
+        f.name
+        for f in schd.data_store_mgr.added[FAMILY_PROXIES].values()
+        if f.name == 'BAR'
+    }) == 1
+    for itask in schd.pool.get_tasks():
+        if itask.tdef.name == 'bar':
+            schd.pool.remove(itask, 'Test removal')
+    schd.data_store_mgr.update_data_structure()
+    assert len({
+        f.name
+        for f in schd.data_store_mgr.added[FAMILY_PROXIES].values()
+        if f.name == 'BAR'
+    }) == 0
+    assert len({
+        f.name
+        for f in data[FAMILY_PROXIES].values()
+        if f.name == 'BAR'
+    }) == 0
+
+
+async def test_family_ascent_point_prune(harness_prune):
```

Comment:
Note that this doesn't fail on 8.4.x - were you expecting it to?

Reply:
There is a [safety catch](https://github.com/dwsutherland/cylc-flow/blob/90f69e9615c765690321e110626afa1c335448a4/cylc/flow/data_store_mgr.py#L1929) in the family pruning method to make sure it doesn't trip up if the family proxy is not found in the added or data store..
Theoretically, this should never happen and out tests didn't trigger it.. But in practice I don't want to exclude some exotic situations..

This coverage required I test it, so I added this one to do just that..